File: src/C64Basic.java
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.charset.Charset;
import java.io.IOException;
import java.nio.file.StandardOpenOption;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public class C64Basic {

    private static class Tokenizer {

        private final StringBuffer source;
        static Map<String,Byte> keywords = new HashMap<>();
        static Map<String,Byte> keycodes = new HashMap<>();
        static String[] commands;
        byte[] code = new byte[38913]; // 38911 basic byte free!
        private int start = 0;
        private int current = 0;
        private int cidx = 0; // code index
        private int nexti;    // where to put index of next tokenized line
        //private int line = 1;
        private boolean LINENO; // looking for a line number?

        public Tokenizer(String source) {
            this.source = new StringBuffer(source);
            // load the HashMaps
            populateKeywords();
            populateKeycodes();

            code[cidx++] = (byte)(0x01);
            code[cidx++] = (byte)(0x08);
            nexti = cidx;
            //cidx += 2; // leave 2 bytes for address link to next line
        }

        byte[] scanTokens() {
            LINENO = true;
            while (!isAtEnd()) {
                // We are at the beginning of the next lexeme.
                start = current;
                scanToken();
            }
            addToken(0);
            // return only what we need.
            byte[] c = new byte[cidx];
            System.arraycopy(code, 0, c, 0, cidx);
            return c;
        }

        private void scanToken() {
            char c = advance();
            switch (c) {
                case '+' : addToken((byte)0xaa); break;
                case '-' : addToken((byte)0xab); break;
                case '*' : addToken((byte)0xac); break;
                case '/' : addToken((byte)0xad); break;
                case '↑' : case '^': addToken((byte)0xae); break;
                case '>' : addToken((byte)0xb1); break;
                case '=' : addToken((byte)0xb2); break;
                case '<' : addToken((byte)0xb3); break;
                case '?' : addToken((byte)0x99); break; // other form of print

                case '"': string(); break;

                case '\n' : case '\r' :
                    // mark end of line
                    addToken((byte)0);
                    int pos = cidx - 2 + 0x0801;
                    System.out.printf("cidx = 0x%04X, pos = 0x%04x%n", cidx, pos);
                    // replace previous link with new address.
                    code[nexti]   = (byte)(pos & 0xff);
                    code[nexti+1] = (byte)((pos & 0xff00) >> 8);
                    LINENO = true;
                    break; // fixup!

                // if it's not special, bring it along.
                default  :
                    if (isDigit(c) && LINENO) {
                        nexti = cidx;
                        cidx += 2; // leave room for the next pointer

                        number();
                        while(peek()==' ') advance(); // consume space(s) after line number
                        LINENO = false;
                    } else if (isAlpha(c))
                        command();
                    else addToken((byte)petscii(c));
            }
        }

        private void command() {
            //System.out.println("entered command()");
            for ( String c : commands ) {
                //System.out.println(c + " " + source.indexOf(c) + " " + start);
                if (source.indexOf(c, start) == start) {
                    byte op = keywords.get(c);
                    System.out.printf("Found %s, 0x%02x%n", c, op);
                    addToken(op);
                    current = start + c.length(); // adjust line
                    // if it's a REM, special handling to end of line.
                    if (op == (byte) 0x8f) {
                        start = current;
                        while (peek() != '\n') advance();
                        addToken(source.substring(start, current));
                    }
                    return;
                }
            }
            addToken(source.substring(start, current));
        }

        private void number() {
            //System.out.println("entered number()");
            while (isDigit(peek())) advance();
            addToken(Integer.parseInt(source.substring(start, current)));
        }

        private void string() {
            //System.out.println("entered string()");
            while (peek() != '"' && !isAtEnd()) {
                if (peek() == '\n') break; //
                if (peek() == '{')
                    keycode();
                else
                    advance();
            }

            // Maybe the closing ". C64 doesn't care about missing close "
            if (peek() != '\n' && peek() != '\r')
                advance();

            String value = source.substring(start, current);
            addToken(value);
        }

        private void keycode() {
            //System.out.println("entered keycode()");
            int s = current;
            byte kc=0;
            while (peek() != '}' && !isAtEnd()) advance();
            advance(); // consume the }
            String value = source.substring(s, current);
            //System.out.println("found " + value);
            if (keycodes.containsKey(value))
                kc = keycodes.get(value);
            else
                try {
                    kc = (byte)(Integer.parseInt(value.substring(1,value.length()-1)) & 0xff);
                } catch (NumberFormatException e) {
                    System.err.println("Invalid keycode: " + value);
                    System.exit(1);
                }
            //System.out.printf("keycode() value = %s, 0x%02X%n", value, kc);
            char c = (char)(kc & 0xff);
            source.replace(s, current, "" + c);
            current = s;
            advance(); // move past the replaced code
        }

        private char peek() {
            if (isAtEnd()) return '\0';
            return source.charAt(current);
        }

        private static boolean isDigit(char c) {
            return c >= '0' && c <= '9';
        }

        private static boolean isAlpha(char c) {
            return (c >= 'A' && c <= 'Z') || ( c >= 'a' && c < 'z');
        }

        private boolean isAtEnd() {
            return current >= source.length();
        }

        private char advance() {
            return source.charAt(current++);
        }

        private void addToken(byte b) {
            //System.out.printf("Adding byte 0x%02x%n", b);
            code[cidx++] = b;
        }

        private void addToken(int i) {
            //System.out.printf("Adding int 0x%04x%n", i);
            code[cidx++] = (byte)(i & 0xff);
            code[cidx++] = (byte)((i & 0xff00) >> 8);
        }

        private void addToken(String s) {
            //System.out.printf("Adding string %s%n", s);
            for (int x = 0; x < s.length(); x++)
                code[cidx++] = (byte)petscii(s.charAt(x));
            //code[cidx++] = (byte)s.charAt(x);
        }

        private char petscii(char c) {
            if ( c >= 'A' && c <= 'Z' ) return (char)(c + 32);
            if ( c >= 'a' && c <= 'z' ) return (char)(c - 32);
            return c;
        }

        private static void populateKeycodes() {
            keycodes.put("{white}", (byte)0x05);
            keycodes.put("{return}", (byte)0x0d);
            keycodes.put("{down}", (byte)0x11);
            keycodes.put("{reverse on}", (byte)0x12);
            keycodes.put("{home}", (byte)0x13);
            keycodes.put("{delete}", (byte)0x14);
            keycodes.put("{red}", (byte)0x1c);
            keycodes.put("{right}", (byte)0x1d);
            keycodes.put("{green}", (byte)0x1e);
            keycodes.put("{blue}", (byte)0x1f);
            keycodes.put("{pound}", (byte)0x5c);
            keycodes.put("{orange}", (byte)0x81);
            keycodes.put("{black}", (byte)0x90);
            keycodes.put("{up}", (byte)0x91);
            keycodes.put("{reverse off}", (byte)0x92);
            keycodes.put("{clear}", (byte)0x93);
            keycodes.put("{brown}", (byte)0x95);
            keycodes.put("{pink}", (byte)0x96);
            keycodes.put("{dark gray}", (byte)0x97);
            keycodes.put("{grey}", (byte)0x98);
            keycodes.put("{light green}", (byte)0x99);
            keycodes.put("{light blue}", (byte)0x9a);
            keycodes.put("{light gray}", (byte)0x9b);
            keycodes.put("{purple}", (byte)0x9c);
            keycodes.put("{left}", (byte)0x9d);
            keycodes.put("{yellow}", (byte)0x9e);
            keycodes.put("{cyan}", (byte)0x9f);
            keycodes.put("{pi}", (byte)0xff);
        }

        private static void populateKeywords(){
            keywords.put("end",(byte)0x80);
            keywords.put("for",(byte)0x81);
            keywords.put("next",(byte)0x82);
            keywords.put("data",(byte)0x83);
            keywords.put("input#",(byte)0x84);
            keywords.put("input",(byte)0x85);
            keywords.put("dim",(byte)0x86);
            keywords.put("read",(byte)0x87);
            keywords.put("let",(byte)0x88);
            keywords.put("goto",(byte)0x89);
            keywords.put("run",(byte)0x8a);
            keywords.put("if",(byte)0x8b);
            keywords.put("restore",(byte)0x8c);
            keywords.put("gosub",(byte)0x8d);
            keywords.put("return",(byte)0x8e);
            keywords.put("rem",(byte)0x8f);
            keywords.put("stop",(byte)0x90);
            keywords.put("on",(byte)0x91);
            keywords.put("wait",(byte)0x92);
            keywords.put("load",(byte)0x93);
            keywords.put("save",(byte)0x94);
            keywords.put("verify",(byte)0x95);
            keywords.put("def",(byte)0x96);
            keywords.put("poke",(byte)0x97);
            keywords.put("print#",(byte)0x98);
            keywords.put("print",(byte)0x99);
            keywords.put("cont",(byte)0x9a);
            keywords.put("list",(byte)0x9b);
            keywords.put("clr",(byte)0x9c);
            keywords.put("cmd",(byte)0x9d);
            keywords.put("sys",(byte)0x9e);
            keywords.put("open",(byte)0x9f);
            keywords.put("close",(byte)0xa0);
            keywords.put("get",(byte)0xa1);
            keywords.put("new",(byte)0xa2);
            keywords.put("tab(",(byte)0xa3);
            keywords.put("to",(byte)0xa4);
            keywords.put("fn",(byte)0xa5);
            keywords.put("spc(",(byte)0xa6);
            keywords.put("then",(byte)0xa7);
            keywords.put("not",(byte)0xa8);
            keywords.put("step",(byte)0xa9);
        /*
        keywords.put("+",(byte)0xaa);
        keywords.put("-",(byte)0xab);
        keywords.put("*",(byte)0xac);
        keywords.put("/",(byte)0xad);
        keywords.put("↑",(byte)0xae);
        */
            keywords.put("and",(byte)0xaf);
            keywords.put("or",(byte)0xb0);
        /*
        keywords.put(">",(byte)0xb1);
        keywords.put("=",(byte)0xb2);
        keywords.put("<",(byte)0xb3);
        */
            keywords.put("sgn",(byte)0xb4);
            keywords.put("int",(byte)0xb5);
            keywords.put("abs",(byte)0xb6);
            keywords.put("usr",(byte)0xb7);
            keywords.put("fre",(byte)0xb7);
            keywords.put("pos",(byte)0xb9);
            keywords.put("sqr",(byte)0xba);
            keywords.put("rnd",(byte)0xbb);
            keywords.put("log",(byte)0xbc);
            keywords.put("exp",(byte)0xbd);
            keywords.put("cos",(byte)0xbe);
            keywords.put("sin",(byte)0xbf);
            keywords.put("tan",(byte)0xc0);
            keywords.put("atn",(byte)0xc1);
            keywords.put("peek",(byte)0xc2);
            keywords.put("len",(byte)0xc3);
            keywords.put("str$",(byte)0xc4);
            keywords.put("val",(byte)0xc5);
            keywords.put("asc",(byte)0xc6);
            keywords.put("chr$",(byte)0xc7);
            keywords.put("left$",(byte)0xc8);
            keywords.put("right$",(byte)0xc9);
            keywords.put("mid$",(byte)0xca);
            keywords.put("go",(byte)0xcb);

            commands = keywords.keySet().toArray(new String[1]);
            Arrays.sort(commands, Collections.reverseOrder());
        }
    }

    private static void hexdump(byte[] bytes) {
        int start = (bytes[1] & 0xff) * 256 + (bytes[0] & 0xff);
        int pc = start;
        StringBuilder chars = new StringBuilder();
        int zeroes = 0;

        for (int x = 2; x < bytes.length; x++) {
            if ( (pc - start) % 8 == 0 ) {
                System.out.printf(" %s", chars);
                if ( zeroes >= 3 ) return;
                System.out.printf("\n%04X:", pc);
                chars.setLength(0);
            }
            char c = (char)(bytes[x] & 0xff);
            zeroes = c==0 ? zeroes + 1 : 0;
            if ( c >= 32 && c <= 127 )
                chars.append(c);
            else
                chars.append('.');
            System.out.printf(" %02X", bytes[x]);
            pc++;

        }
        System.out.println("\n");
    }

    public static void main(String[] args) throws IOException {
        byte[] bytes;

        if (args.length != 2) {
            System.out.println("Must provide source and object.");
            System.exit(3);
        }

        bytes = Files.readAllBytes(Paths.get(args[0]));

        Tokenizer tzr = new Tokenizer(new String(bytes, Charset.defaultCharset()));
        byte[] program = tzr.scanTokens();

        hexdump(program);
        Files.write(Paths.get(args[1]), program, StandardOpenOption.CREATE);
    }
}

File: src/C64Basic.c

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <errno.h>
#include "git-banned.h"

static unsigned char *scanTokens(int *len);

static void scanToken();

static void command();

static void number();

static void string();

static char keycode();

static char peek();

static int isDigit(char c);

static int isAlpha(char c);

static int isAtEnd();

static char advance();

static void addByteToken(char b);

static void addIntToken(int i);

static void addStrToken(const char *s);

static char petscii(char c);

static long myatol(const char *buf);

static void hexDump(const unsigned char *bytes, long length);

static void writeProgram(const char *fname, const unsigned char *program, int len);

static char *readAllBytes(const char *fname, long *len);

// Yeah, there's a lot of globals... so?
char *source;
int srcLen;
int start = 0; // start of the code section being tokenized
int current = 0; // current position >= start
int cidx = 0; // code index
int nexti; // where to put index of next tokenized line
unsigned char code[38913]; // 38911 basic byte free!
bool LINENO; // looking for a line number?

// keyword/keycode struct.
typedef struct {
    char *name;
    unsigned char val;
} Keyword;

// Set keywords and calculate length
Keyword keywords[] = {
    {"end", 0x80},
    {"for", 0x81},
    {"next", 0x82},
    {"data", 0x83},
    {"input#", 0x84},
    {"input", 0x85},
    {"dim", 0x86},
    {"read", 0x87},
    {"let", 0x88},
    {"goto", 0x89},
    {"run", 0x8a},
    {"if", 0x8b},
    {"restore", 0x8c},
    {"gosub", 0x8d},
    {"return", 0x8e},
    {"rem", 0x8f},
    {"stop", 0x90},
    {"on", 0x91},
    {"wait", 0x92},
    {"load", 0x93},
    {"save", 0x94},
    {"verify", 0x95},
    {"def", 0x96},
    {"poke", 0x97},
    {"print#", 0x98},
    {"print", 0x99},
    {"cont", 0x9a},
    {"list", 0x9b},
    {"clr", 0x9c},
    {"cmd", 0x9d},
    {"sys", 0x9e},
    {"open", 0x9f},
    {"close", 0xa0},
    {"get", 0xa1},
    {"new", 0xa2},
    {"tab(", 0xa3},
    {"to", 0xa4},
    {"fn", 0xa5},
    {"spc(", 0xa6},
    {"then", 0xa7},
    {"not", 0xa8},
    {"step", 0xa9},
    /*
    {"+",0xaa},
    {"-",0xab},
    {"*",0xac},
    {"/",0xad},
    {"↑",0xae},
    */
    {"and", 0xaf},
    {"or", 0xb0},
    /*
    {">",0xb1},
    {"=",0xb2},
    {"<",0xb3},
    */
    {"sgn", 0xb4},
    {"int", 0xb5},
    {"abs", 0xb6},
    {"usr", 0xb7},
    {"fre", 0xb8},
    {"pos", 0xb9},
    {"sqr", 0xba},
    {"rnd", 0xbb},
    {"log", 0xbc},
    {"exp", 0xbd},
    {"cos", 0xbe},
    {"sin", 0xbf},
    {"tan", 0xc0},
    {"atn", 0xc1},
    {"peek", 0xc2},
    {"len", 0xc3},
    {"str$", 0xc4},
    {"val", 0xc5},
    {"asc", 0xc6},
    {"chr$", 0xc7},
    {"left$", 0xc8},
    {"right$", 0xc9},
    {"mid$", 0xca},
    {"go", 0xcb}
};
int keywordLength = sizeof(keywords) / sizeof(keywords[0]);

// Set keycodes and calculate length
Keyword keycodes[] = {
    {"{white}", 0x05},
    {"{return}", 0x0d},
    {"{down}", 0x11},
    {"{reverse on}", 0x12},
    {"{home}", 0x13},
    {"{delete}", 0x14},
    {"{red}", 0x1c},
    {"{right}", 0x1d},
    {"{green}", 0x1e},
    {"{blue}", 0x1f},
    {"{pound}", 0x5c},
    {"{orange}", 0x81},
    {"{black}", 0x90},
    {"{up}", 0x91},
    {"{reverse off}", 0x92},
    {"{clear}", 0x93},
    {"{brown}", 0x95},
    {"{pink}", 0x96},
    {"{dark gray}", 0x97},
    {"{grey}", 0x98},
    {"{light green}", 0x99},
    {"{light blue}", 0x9a},
    {"{light gray}", 0x9b},
    {"{purple}", 0x9c},
    {"{left}", 0x9d},
    {"{yellow}", 0x9e},
    {"{cyan}", 0x9f},
    {"{pi}", 0xff}
};
int keycodeLength = sizeof(keycodes) / sizeof(keycodes[0]);


// Kickoff the token scanning!
static unsigned char *scanTokens(int *len) {
    // First two bytes are the starting address on C64
    code[cidx++] = 0x01;
    code[cidx++] = 0x08;
    nexti = cidx;
    //cidx += 2; // leave 2 bytes for address link to next line

    LINENO = true;
    while (!isAtEnd()) {
        // We are at the beginning of the next lexeme.
        start = current;
        scanToken();
    }
    // At this point, all source code has been processed.
    addIntToken(0); // null pointer at end of program.
    // return only what we need.
    *len = cidx;
    return code;
}

// Find simple tokens and predict larger ones.
static void scanToken() {
    char c = advance();
    switch (c) {
        case '+': addByteToken((char) 0xaa);
            break;
        case '-': addByteToken((char) 0xab);
            break;
        case '*': addByteToken((char) 0xac);
            break;
        case '/': addByteToken((char) 0xad);
            break;
        /*case '↑' : */
        case '^': addByteToken((char) 0xae);
            break;
        case '>': addByteToken((char) 0xb1);
            break;
        case '=': addByteToken((char) 0xb2);
            break;
        case '<': addByteToken((char) 0xb3);
            break;
        case '?': addByteToken((char) 0x99);
            break; // other form of print

        case '"': string();
            break;

        case '\n':
        case '\r':
            // mark end of tokenized line
            addByteToken(0);
            int pos = cidx - 2 + 0x0801;
            //printf("cidx = 0x%04X, pos = 0x%04x\n", cidx, pos);
            // replace previous link with new address.
            code[nexti] = (char) (pos & 0xff);
            code[nexti + 1] = (char) ((pos & 0xff00) >> 8);
            LINENO = true;
            break; // fixup!

        // if it's not special, bring it along.
        default:
            if (isDigit(c) && LINENO) {
                nexti = cidx;
                cidx += 2; // leave room for the next pointer
                number();
                while (peek() == ' ') advance(); // consume space(s) after line number
                LINENO = false;
            } else if (isAlpha(c))
                command();
            else addByteToken(petscii(c));
    }
}

// Build a command token
static void command() {
    //System.out.println("entered command()");
    for (int x = 0; x < keywordLength; x++) {
        const char *c = keywords[x].name;
        if (strstr(source + start, c) == source + start) {
            unsigned char op = keywords[x].val;
            //printf("Found %s, 0x%02x\n", c, op);
            addByteToken(op);
            current = start + (int) strlen(c); // adjust line
            // if it's a REM, special handling to end of line.
            if (op == 0x8f) {
                start = current;
                while (peek() != '\n') advance();
                char *s = strndup(source + start, current - start);
                addStrToken(s);
                free(s);
            }
            return; // matched a keyword, leave.
        }
    }
    // Add raw text that didn't match
    char *s = strndup(source + start, current - start);
    addStrToken(s);
    free(s);
}

// Build a number token
static void number() {
    //System.out.println("entered number()");
    while (isDigit(peek())) advance();
    char *s = strndup(source + start, current - start);
    addIntToken((int) myatol(s));
    free(s);
}

// Build a string token
static void string() {
    //System.out.println("entered string()");
    // Use a local string limited to 255 characters.
    // This is the limit of a single C64 BASIC tokenized line, so...
    char bs[256] = {0};
    int bsp = 0;
    bs[bsp++] = '"'; // remember we already saw this...

    char c = peek();
    // This is different from the Java version because we can't just replace the substring.
    while (c != '"' && !isAtEnd() && bsp < 255) {
        if (c == '\n') break; //
        if (c == '{') {
            // keycode handling is just...messy.
            char kc = keycode();
            bs[bsp++] = kc;
        } else {
            advance();
            bs[bsp++] = c;
        }
        c = peek();
    }
    bs[bsp] = 0; // end the string! possibly redundant.

    // Maybe the closing ". C64 doesn't care about missing close "
    c = peek();
    if (c != '\n' && c != '\r') {
        bs[bsp] = c;
        advance();
    }
    addStrToken(bs);
}

static char keycode() {
    //System.out.println("entered keycode()");
    int s = current;
    unsigned char kc = 0;
    while (peek() != '}' && !isAtEnd()) advance();
    advance(); // consume the }
    char *value = strndup(source + s, current - s);
    printf("%s\n", value);
    //System.out.println("found " + value);
    for (int x = 0; x < keycodeLength; x++) {
        if (strcmp(keycodes[x].name, value) == 0) {
            kc = keycodes[x].val;
            break;
        }
    }
    // Keycodes without words can be decimal values
    if (kc == 0) {
        kc = (char) (myatol(value + 1) & 0xff);
    }
    //System.out.printf("keycode() value = %s, 0x%02X%n", value, kc);
    free(value);
    return kc;
    //current = s;
}

// Look ahead.
static char peek() {
    if (isAtEnd()) return '\0';
    return source[current];
}

static int isDigit(char c) {
    return c >= '0' && c <= '9';
}

static int isAlpha(char c) {
    return (c >= 'A' && c <= 'Z') || (c >= 'a' && c < 'z');
}

static int isAtEnd() {
    return current >= srcLen;
}

static char advance() {
    return source[current++];
}

// Add token as bvte
static void addByteToken(char b) {
    //System.out.printf("Adding byte 0x%02x%n", b);
    code[cidx++] = b;
}

// Add token as int - little endian
static void addIntToken(int i) {
    //System.out.printf("Adding int 0x%04x%n", i);
    code[cidx++] = (char) (i & 0xff);
    code[cidx++] = (char) ((i & 0xff00) >> 8);
}

// Add token as string
static void addStrToken(const char *s) {
    //System.out.printf("Adding string %s%n", s);
    size_t len = strlen(s);
    for (int x = 0; x < len; x++)
        code[cidx++] = petscii(s[x]);
    //code[cidx++] = (byte)s.charAt(x);
}

// Convert to PETSCII
static char petscii(char c) {
    if (c >= 'A' && c <= 'Z') return (char) (c + 32);
    if (c >= 'a' && c <= 'z') return (char) (c - 32);
    return c;
}

// Safe atol()
static long myatol(const char *buf) {
    errno = 0;
    char *p;
    long a = strtol(buf, &p, 10); // also sets ERANGE

    // *p can be '\0' or '\n', but p cannot be buf.
    if (!((!*p || *p == '\n') && p != buf && !errno))
        errno = EINVAL;
    return a;
}

// A Sample hexdump (https://programmingby.design/algorithms/the-hex-dump/)
static void hexDump(const unsigned char *bytes, long length) {
    // we and with 0xff to mask off the bits when sign-extended.
    int st = bytes[1] * 256 + bytes[0];
    printf("%d\n", st);
    int pc = st, c = 0, x;
    char chars[10] = {0};

    for (x = 2; x < length; x++) {
        // make it pretty
        if ((pc - st) % 8 == 0) {
            // chars are built during each iteration, but
            // printed when we reach the end of the line.
            printf(" %s", chars);
            printf("\n%04X:", pc);
            c = 0;
        }
        unsigned char ch = bytes[x];
        // build the chars - only printable chars
        chars[c++] = (ch >= 32 && ch <= 127) ? ch : '.';

        printf(" %02X", bytes[x]);
        pc++;
    }

    // fix final row
    int last = (pc - st) % 8;
    chars[c] = '\0';
    if (last > 0)
        for (x = 0; x < 8 - last; x++)
            printf("   ");
    printf(" %s\n", chars);
}

static void writeProgram(const char *fname, const unsigned char *program, int len) {
    FILE *file = fopen(fname, "w");
    if (file == NULL) {
        perror("Error opening object file");
        return;
    }
    size_t bytesWritten = fwrite(program, 1, len, file);
    if (bytesWritten != len)
        perror("Error writing object file");
    fclose(file);
}

// Read a complete file into memory.
static char *readAllBytes(const char *fname, long *len) {
    // open file
    FILE *file = fopen(fname, "r");
    if (file == NULL) {
        perror("Error opening source file");
        return NULL;
    }

    // get size
    fseek(file, 0, SEEK_END);
    long fileSize = ftell(file);
    rewind(file);

    // get the mem
    char *buffer = (char *) malloc(fileSize * sizeof(char) + 1);

    // Read the entire file into the buffer
    size_t bytesRead = fread(buffer, 1, fileSize, file);
    if (bytesRead != fileSize) {
        perror("Error reading file");
        fclose(file);
        free(buffer);
        *len = 0;
        return NULL;
    }
    // add the terminator
    buffer[fileSize] = '\0';
    *len = fileSize; // be sure to note the size!
    return buffer;
}

int main(int argc, char **argv) {
    // Check for proper number of arguments
    if (argc != 3) {
        printf("\nMust provide source and object.\n");
        exit(3);
    }

    printf("%s\n", argv[1]);
    // Read the file.
    long sl;
    source = readAllBytes(argv[1], &sl);
    srcLen = (int) sl;

    // Send to Tokenizer as String.
    int pl;
    unsigned char *program = scanTokens(&pl);
    free(source);

    // Show hexdump and write the resultant program.
    hexDump(program, pl);
    writeProgram(argv[2], program, pl);
}
